File: Trigger/TrigHypothesis/TrigLongLivedParticlesHypo/src/TrigIsoHPtTrackTriggerHypoTool.cxx
/*
  Copyright (C) 2002-2021 CERN for the benefit of the ATLAS collaboration

  * Trigger Hypo Tool, that is aimed at triggering high pt isolated tracks 
  * author Ismet Siral <ismet.siral@cern.ch> - University of Oregon
*/


#include "TrigCompositeUtils/HLTIdentifier.h"
#include "TrigCompositeUtils/Combinators.h"
#include "AthenaMonitoringKernel/Monitored.h"
#include "TrigIsoHPtTrackTriggerHypoTool.h"
#include "GaudiKernel/PhysicalConstants.h"

using namespace TrigCompositeUtils;

TrigIsoHPtTrackTriggerHypoTool::TrigIsoHPtTrackTriggerHypoTool( const std::string& type, 
						const std::string& name, 
						const IInterface* parent ) 
  : AthAlgTool( type, name, parent ),
    m_decisionId( HLT::Identifier::fromToolName( name ) ) {}

StatusCode TrigIsoHPtTrackTriggerHypoTool::initialize()  {
  
  if ( !m_monTool.empty() ) CHECK( m_monTool.retrieve() );

  ATH_MSG_VERBOSE( "Initialization completed successfully:" );

  //Track Trigger Kinematic Requirments
  ATH_MSG_VERBOSE( "MaxTrackPt           = " << m_TrackPt ); 
  ATH_MSG_VERBOSE( "MaxTrackEta          = " << m_TrackEta ); 
  ATH_MSG_VERBOSE( "MaxTrackd0           = " << m_Trackd0 );
  ATH_MSG_VERBOSE( "MaxTrackd0Sig        = " << m_Trackd0Sig );
  ATH_MSG_VERBOSE( "MinTrackNPixHits     = " << m_TrackNPixHits );
  ATH_MSG_VERBOSE( "MinTrackNSCTHits     = " << m_TrackNSCTHits );
  //Track Trigger Isolation Requirments
  ATH_MSG_VERBOSE( "TrackIsoCone         = " << m_IsoDR );
  ATH_MSG_VERBOSE( "MinIsoTrackPt        = " << m_IsoPt );
  ATH_MSG_VERBOSE( "EnableTrackIsolation = " << m_doIso );
  ATH_MSG_VERBOSE( "EnableCumulIsolation = " << m_IsoCum );

  std::vector<size_t> sizes( {m_TrackPt.size(), m_TrackEta.size(), m_Trackd0.size( ), m_Trackd0Sig.size(), m_TrackNPixHits.size() ,  m_TrackNSCTHits.size() , m_doIso.size() , m_IsoDR.size() , m_IsoPt.size() } );


  if ( *std::min_element( sizes.begin(), sizes.end() ) != *std::max_element( sizes.begin(), sizes.end() )  ) {
    ATH_MSG_ERROR( "Missconfiguration, cut properties listed above ( when INFO ) have different dimensions shortest: " <<  *std::min_element( sizes.begin(), sizes.end() ) << " longest " << *std::max_element( sizes.begin(), sizes.end() ) );
    return StatusCode::FAILURE;
  }

  ATH_MSG_VERBOSE( "Tool configured for chain/id: " << m_decisionId  );

  return StatusCode::SUCCESS;
}


TrigIsoHPtTrackTriggerHypoTool::~TrigIsoHPtTrackTriggerHypoTool() {}

bool TrigIsoHPtTrackTriggerHypoTool::decideOnSingleObject( const xAOD::TrackParticle_v1* track, const xAOD::TrackParticleContainer* AllTracks, size_t cutIndex ) const {
  auto cutCounter = Monitored::Scalar<int>( "CutCounter", -1 );  
  auto cutIndexM  = Monitored::Scalar<int>( "CutIndex", cutIndex );  // one can do 2D plots for each cut independently

  auto MONtrackPt     = Monitored::Scalar( "trackPt", -999. );
  auto MONtrackd0    = Monitored::Scalar( "trackd0", -999. );
  auto MONtrackNPixHits   = Monitored::Scalar( "trackNPixHits", -999. );
  auto MONtrackNSCTHits   = Monitored::Scalar( "trackNSCTHits", -999. );
  auto MONtrackd0Sig   = Monitored::Scalar( "trackd0Sig", -999. );
  auto MONtrackEta     = Monitored::Scalar( "trackEta", -999. );
  auto MONtrackIsoPt   = Monitored::Scalar( "trackIsoPt", -999. );
  auto MONtrackAggrIsoPt   = Monitored::Scalar( "trackAggrIsoPt", -999. );
  auto monitorIt  = Monitored::Group( m_monTool, cutCounter, cutIndexM,
				      MONtrackPt, MONtrackEta, 
				      MONtrackd0, MONtrackd0Sig,
				      MONtrackNPixHits, MONtrackNSCTHits, 
				      MONtrackIsoPt, MONtrackAggrIsoPt   
				      );




  //Checking Track pT Requirments
  auto trackPt= (track)->pt();
  if( trackPt < m_TrackPt[cutIndex] ) { // Check track pT requirments 
    ATH_MSG_DEBUG( "Fails pt cut" << trackPt << " < " << m_TrackPt[cutIndex] );
    return false;

  }
  cutCounter++;


  //Checking Track Eta Requirments
  auto trackEta=std::abs( (track)->p4().Eta() );
  if(  trackEta > m_TrackEta[cutIndex] ) { // Check track pT requirments 
    ATH_MSG_DEBUG( "Fails Eta cut" << trackEta << " > " << m_TrackEta[cutIndex] );
    return false;

  }
  cutCounter++;


  
  //Checking Track d0 Requirments
  auto trackd0= std::abs((track)->d0());
  if( trackd0 > m_Trackd0[cutIndex]) {
    ATH_MSG_DEBUG( "Fails d0 cut" << trackd0 << " > " << m_Trackd0[cutIndex] );
    return false;

  }
  cutCounter++;


  //Checking Track d0 Sig Requirments
  auto trackd0Sig= std::abs(xAOD::TrackingHelpers::d0significance(track));
  if( trackd0Sig > m_Trackd0Sig[cutIndex]) {
    ATH_MSG_DEBUG( "Fails d0 Sig cut" << trackd0Sig << " > " << m_Trackd0Sig[cutIndex] );
    return false;

  }
  cutCounter++;


  //Attempt to read track summary, and save info on number of pixels
  uint8_t trackNPixHits=0;
  if ( ! (track)->summaryValue(trackNPixHits,xAOD::numberOfPixelHits )) { // Cannot obtain the track summary
    ATH_MSG_DEBUG( "Failed to retrieve pedigree parameters");  
    return  false;
  }
  ATH_MSG_DEBUG( "Succesfully retrieved pedigree parameters");  
  cutCounter++;
 //Attempt to read track summary, and save info on number of SCT hits
  uint8_t trackNSCTHits=0;
  if ( ! (track)->summaryValue(trackNSCTHits,xAOD::numberOfSCTHits )) { // Cannot obtain the track summary
    ATH_MSG_DEBUG( "Failed to retrieve pedigree parameters");  
    return  false;
  }
  ATH_MSG_DEBUG( "Succesfully retrieved pedigree parameters");  
  cutCounter++;



  //Checking Track Min number of Pix Hits Requirments
  if( trackNPixHits <= m_TrackNPixHits[cutIndex] ){

    ATH_MSG_DEBUG( "Fails numperOfPixelHits cut" << trackNPixHits  << " <= " << m_TrackNPixHits[cutIndex]  );
    return false;
  }
  cutCounter++;    


  //Checking Track Min number of SCT Hits Requirments
  if( trackNSCTHits <= m_TrackNSCTHits[cutIndex] ){

    ATH_MSG_DEBUG( "Fails numperOfSCTHits cut" << trackNSCTHits  << " <= " << m_TrackNSCTHits[cutIndex]  );
    return false;
  }
  cutCounter++;    


  //Definning a cumlative pT variable that adds up the momentum of all tracks

  //If isolation is applied, loop over all tracks, and veto the event if there is a track above a certain threshold in the isolation cone
  if (m_doIso[cutIndex]) {
    float CumulativePT = 0.0;
    
    for (auto trackIter = AllTracks->begin(); trackIter != AllTracks->end(); ++trackIter){

      //Skip the track that is out of DR
      if( (*trackIter)==track) continue;
      if (track->p4().DeltaR((*trackIter)->p4()) > m_IsoDR[cutIndex] ) continue;

      //Skip the track that doens't have many Pixel and SCT Hits (Quality Check)
      uint8_t iterPix=0;
      if ( ! (*trackIter)->summaryValue(iterPix,xAOD::numberOfPixelHits ))  
	continue;
      uint8_t iterSCT=0;
      if ( ! (*trackIter)->summaryValue(iterSCT,xAOD::numberOfSCTHits )) 
	continue;

      if (iterPix<=m_TrackNPixHits[cutIndex] ) continue;
      if (iterSCT<=2 ) continue;


      //If cumalitve, add up the momentum of the track, if it's a bove a certain threshold kill the trac

      if(m_IsoCum[cutIndex]){
	CumulativePT+=(*trackIter)->pt();
	if(CumulativePT>=m_IsoPt[cutIndex]) {
	  MONtrackAggrIsoPt=CumulativePT;
	  ATH_MSG_DEBUG( "Fails Cum Isolation cut"  );    
	  return false;
	}
      }
      //If not cumlaitve, veto the track only if one track is bigger then the threshold
      else {
	if ((*trackIter)->pt() > MONtrackIsoPt ) {
	  MONtrackIsoPt=(*trackIter)->pt();
	}

	if( (*trackIter)->pt()>=m_IsoPt[cutIndex] ) {
	  ATH_MSG_DEBUG( "Fails Isolation cut"  );    
	  return false;
	}
      }


     
    }

    MONtrackAggrIsoPt=CumulativePT;
    cutCounter++;
  }

  //Monitorung histograms are filled at the end, only if they pass the selection. (For optimisation studies)
  MONtrackNPixHits=trackNPixHits;
  MONtrackNSCTHits=trackNSCTHits; 
  MONtrackPt= trackPt;
  MONtrackEta= trackEta;
  MONtrackd0= trackd0;
  MONtrackd0Sig= trackd0Sig;



  ATH_MSG_DEBUG( "Passed selection" );
  return  true;

}

StatusCode TrigIsoHPtTrackTriggerHypoTool::inclusiveSelection( std::vector<TrackInfo>& input ) const {
    for ( auto i: input ) {

      auto objDecision = decideOnSingleObject( i.track, i.AllTracks, 0 );
      if ( objDecision == true ) {

	addDecisionID( m_decisionId.numeric(), i.decision  );
      }
    }
    return StatusCode::SUCCESS;
}


StatusCode TrigIsoHPtTrackTriggerHypoTool::markPassing( std::vector<TrackInfo>& input, const std::set<size_t>& passing ) const {

  for ( auto idx: passing ) 
    addDecisionID( m_decisionId.numeric(), input[idx].decision );
  return StatusCode::SUCCESS;

}


StatusCode TrigIsoHPtTrackTriggerHypoTool::multiplicitySelection( std::vector<TrackInfo>& input ) const {
  HLT::Index2DVec passingSelection( m_multiplicity );
  
  for ( size_t cutIndex = 0; cutIndex < m_multiplicity; ++ cutIndex ) {
    size_t trkIndex{ 0 };
    for ( auto trkIter =  input.begin(); trkIter != input.end(); ++trkIter, ++trkIndex ) {
      
      if ( decideOnSingleObject( trkIter->track, trkIter->AllTracks, cutIndex ) ) {
	  passingSelection[cutIndex].push_back( trkIndex );

      }
    }
    // checking if by chance none of the objects passed the single obj selection, if so there will be no valid combination and we can skip
    if ( passingSelection[cutIndex].empty() ) {
      ATH_MSG_DEBUG( "No object passed selection " << cutIndex << " rejecting" );
      return StatusCode::SUCCESS;
    }
  }

  std::set<size_t> passingIndices;
  HLT::elementsInUniqueCombinations( passingSelection, passingIndices );
  return markPassing( input, passingIndices );
  

}

StatusCode TrigIsoHPtTrackTriggerHypoTool::decide(  std::vector<TrackInfo>& input )  const {
  // handle the simplest and most common case ( multiplicity == 1 ) in easiest possible manner
  if ( m_TrackPt.size() == 1 ) {
    return inclusiveSelection( input );

  } else {    
    return multiplicitySelection( input );    
  }

  return StatusCode::SUCCESS;
}

File: Trigger/TrigHypothesis/TrigLongLivedParticlesHypo/python/TrigIsoHPtTrackTriggerHypoTool.py
# Copyright (C) 2002-2021 CERN for the benefit of the ATLAS collaboration

from AthenaCommon.SystemOfUnits import GeV
from AthenaCommon.Logging import logging
# from AthenaCommon.Include import Include 
log = logging.getLogger("TrigLongLivedParticlesHypo.TrigIsoHPtTrackTriggerHypoTool")
from AthenaMonitoringKernel.GenericMonitoringTool import GenericMonitoringTool



log = logging.getLogger('TrigIsoHPtTrackTriggerHypoTool')

def TrigIsoHPtTrackTriggerHypoToolFromDict( chainDict ):
    """ Use menu decoded chain dictionary to configure the tool """
    cparts = [i for i in chainDict['chainParts'] if i['signature']=='UnconventionalTracking']
    
    thresholds = sum([ [cpart['threshold']]*int(cpart['multiplicity']) for cpart in cparts], [])

    name = chainDict['chainName']
    from AthenaConfiguration.ComponentFactory import CompFactory
    tool = CompFactory.TrigIsoHPtTrackTriggerHypoTool(name)

    monTool = GenericMonitoringTool("IM_MonTool"+name)
    monTool.defineHistogram('CutCounter', type='TH1I', path='EXPERT', title="IsoHPtTrackTrigger Hypo Cut Counter;Cut Counter", xbins=8, xmin=-1.5, xmax=7.5, opt="kCumulative") 
    monTool.defineHistogram('trackPt', type='TH1F', path='EXPERT', title="IsoHPtTrackTrigger Hypo p_{T}^{track} [MeV];p_{T}^{track} [MeV];Nevents", xbins=30, xmin=0, xmax=300000) 
    monTool.defineHistogram('trackd0', type='TH1F', path='EXPERT', title="IsoHPtTrackTrigger Hypo d0^{track} ; d0^{track} ;Nevents", xbins=100, xmin=0, xmax=10)
    monTool.defineHistogram('trackNPixHits', type='TH1F', path='EXPERT', title="IsoHPtTrackTrigger Hypo N. Pix Hits^{track} ; N. Pix Hits^{track} ;Nevents", xbins=10, xmin=0, xmax=10) 

    monTool.defineHistogram('trackNSCTHits', type='TH1F', path='EXPERT', title="IsoHPtTrackTrigger Hypo N. SCT Hits^{track} ; N. SCT Hits^{track} ;Nevents", xbins=20, xmin=0, xmax=20) 
    monTool.defineHistogram('trackd0Sig', type='TH1F', path='EXPERT', title="IsoHPtTrackTrigger Hypo d0 Sig^{track} ; d0 Sig^{track} ;Nevents", xbins=60, xmin=0, xmax=6)
    monTool.defineHistogram('trackEta', type='TH1F', path='EXPERT', title="IsoHPtTrackTrigger Hypo Eta^{track} ; Eta^{track} ;Nevents", xbins=30, xmin=-3, xmax=3) 
    monTool.defineHistogram('trackIsoPt', type='TH1F', path='EXPERT', title="IsoHPtTrackTrigger Hypo Iso p_{T}^{track} [MeV]; Isolation Cone p_{T}^{track} [MeV];Nevents", xbins=100, xmin=0, xmax=100000) 
    monTool.defineHistogram('trackAggrIsoPt', type='TH1F', path='EXPERT', title="IsoHPtTrackTrigger Hypo Iso p_{T}^{track} [MeV]; Isolation Cumalitive Cone p_{T}^{track} [MeV];Nevents", xbins=100, xmin=0, xmax=100000) 



    monTool.HistPath = 'IsoHPtTrackTriggerHypoAlgMT/'+tool.getName()
    tool.MonTool = monTool

    strThr = ""

    thresholds = [ float(THR)*GeV for THR in thresholds]
    
    for THR in thresholds:
        strThr += str(THR)+", "
        
    log.debug("Threshold Values are: %s",strThr)

    nt = len( thresholds )
    tool.MinTrackPt = thresholds
    tool.MinTrackNPixHits = [ 2 ] *nt
    tool.MinTrackNSCTHits = [ 5 ] *nt

    EnIso=[]
    IsoCum=[]
    IsoCone=[]
    IsoPt=[]

    TrackEta=[]
    Trackd0=[]
    Trackd0Sig=[]


    for cpart in cparts:
        # For isolation working points are defined iloose/medium/aggrmedium/aggrloose  
        log.debug("Isolation is bound to be set to %s and event working point %s",cpart['isoInfo'],cpart['IDinfo'])
        if cpart['isoInfo'] =="iloose":
            log.debug("Loose isolation is set")
            EnIso.append(True)
            IsoCum.append(False)
            IsoCone.append(0.3)
            IsoPt.append(5*GeV)
        elif cpart['isoInfo'] =="imedium":
            log.debug("Medium isolation is set")
            EnIso.append(True)
            IsoCum.append(False)
            IsoCone.append(0.3)
            IsoPt.append(3*GeV)
        elif cpart['isoInfo'] =="iaggrloose":
            log.debug("Loose Cummulative(Aggregate) isolation is set")
            EnIso.append(True)
            IsoCum.append(True)
            IsoCone.append(0.3)
            IsoPt.append(20*GeV)
        elif cpart['isoInfo'] =="iaggrmedium":
            log.debug("Medium Cummulative(Aggregate) isolation is set")
            EnIso.append(True)
            IsoCum.append(True)
            IsoCone.append(0.3)
            IsoPt.append(10*GeV)
        else:
            log.debug("Isolation is not set")
            EnIso.append(False)
            IsoCum.append(False)
            IsoCone.append(0.3)
            IsoPt.append(5*GeV)

        #Working point defenitions
        if cpart['IDinfo'] =="loose":
            log.debug("Loose ID working point is set")
            TrackEta.append(2.5)
            Trackd0.append(1000.)
            Trackd0Sig.append(10.)
            
        elif cpart['IDinfo'] =="tight":
            log.debug("Tight ID working point is set")
            TrackEta.append(2.0)
            Trackd0.append(1000.)
            Trackd0Sig.append(5.)
        else:
            log.debug("Medium ID working point is set")
            TrackEta.append(2.5)
            Trackd0.append(1000.)
            Trackd0Sig.append(5.)



    tool.MinTrackEta = TrackEta
    tool.MaxTrackd0 = Trackd0
    tool.MaxTrackd0Sig = Trackd0Sig
    
    tool.EnableTrackIsolation = EnIso
    tool.EnableCumalitiveIsolation = IsoCum
    tool.TrackIsoCone = IsoCone
    tool.MinIsoTrackPt = IsoPt

    return tool


def TrigIsoHPtTrackTriggerHypoToolFromName( name, conf ):
    """ provides configuration of the hypo tool giben the chain name
    The argument will be replaced by "parsed" chain dict. For now it only serves simplest chain HLT_eXYZ.
    """
    from TriggerMenuMT.HLTMenuConfig.Menu.DictFromChainName import dictFromChainName

    decodedDict = dictFromChainName(conf)
        
    return TrigIsoHPtTrackTriggerHypoToolFromDict( decodedDict )



if __name__ == "__main__":
    tool = TrigIsoHPtTrackTriggerHypoToolFromName("HLT_unconvtrk50_isohpttrack_imedium_L1XE50", "HLT_unconvtrk50_isohpttrack_imedium_L1XE50")
    assert tool, "Not configured simple tool"


    log.debug("ALL OK")
